src/shared/components/base-question-app.tsx

```diff
@@ -0,0 +1,85 @@
+import React, { useRef } from "react";
+import { useLARAInteractiveAPI } from "../hooks/use-lara-interactive-api";
+import { useAutoHeight } from "../hooks/use-auto-height";
+import { useHint } from "../hooks/use-hint";
+import { useRequiredQuestion } from "../hooks/use-required-question";
+import { BaseAuthoring, IBaseAuthoringProps } from "./base-authoring";
+
+import css from "./base-question-app.scss";
+
+interface IBaseQuestionAuthoredState {
+  version: number;
+  hint?: string;
+  required?: boolean;
+}
+
+interface IBaseQuestionInteractiveState {
+  submitted?: boolean;
+}
+
+interface IAuthoringComponentProps<IAuthoredState> {
+  authoredState: IAuthoredState | undefined,
+  setAuthoredState?: (state: IAuthoredState) => void;
+}
+
+interface IRuntimeComponentProps<IAuthoredState, IInteractiveState> {
+  authoredState: IAuthoredState,
+  interactiveState: IInteractiveState | undefined,
+  setInteractiveState?: (state: IInteractiveState) => void;
+  report?: boolean;
+}
+
+interface IProps<IAuthoredState, IInteractiveState> {
+  // Question can provide either fully custom Authoring component or use BaseAuthoring (that uses react-jsonschema-form).
+  Authoring?: React.FC<IAuthoringComponentProps<IAuthoredState>>;
+  baseAuthoringProps?: Omit<IBaseAuthoringProps<IAuthoredState>, "authoredState" | "setAuthoredState">;
+  Runtime: React.FC<IRuntimeComponentProps<IAuthoredState, IInteractiveState>>;
+  isAnswered: (state: IInteractiveState | undefined) => boolean;
+}
+
+export const BaseQuestionApp = <IAuthoredState extends IBaseQuestionAuthoredState, IInteractiveState extends IBaseQuestionInteractiveState>(
+  { Authoring, baseAuthoringProps, Runtime, isAnswered }: IProps<IAuthoredState, IInteractiveState>
+) => {
+  const container = useRef<HTMLDivElement>(null);
+  const { mode, authoredState, interactiveState, setInteractiveState, setAuthoredState, setHeight, setHint, setNavigation } = useLARAInteractiveAPI<IAuthoredState, IInteractiveState>({
+    interactiveState: true,
+    authoredState: true,
+  });
+  useAutoHeight({ container, setHeight });
+  useHint({ authoredState, setHint });
+  const { submitButton, lockedInfo } = useRequiredQuestion({ authoredState, interactiveState, setInteractiveState, setNavigation, isAnswered: isAnswered(interactiveState) });
+
+  const report = mode === "report";
+  const authoring = mode === "authoring";
+  const runtime = mode === "runtime";
+
+  let AuthoringComp = null;
+  if (authoring) {
+    if (Authoring) {
+      AuthoringComp = <Authoring authoredState={authoredState} setAuthoredState={setAuthoredState} />;
+    }
+    if (!Authoring && baseAuthoringProps) {
+      AuthoringComp = <BaseAuthoring {...baseAuthoringProps} authoredState={authoredState} setAuthoredState={setAuthoredState} />;
+    }
+  }
+
+  return (
+    <div ref={container}>
+      { AuthoringComp }
+      {
```

Comment:
This was a little confusing to me since it seems like it is possible to show both the AuthoringComp and the Runtime component at the same time. I realized that isn't possible because the AuthComp will be null if the mode is authoring or runtime. 

Perhaps it is worth putting this in a named method or self invoking method and use switch case like:
```
switch(mode){
case "authoring": 
  if(Authoring)....
case "runtime":
  return <div ><Runtime authoredState={authoredState} interactiveState={interactiveState} setInteractiveState={setInteractiveState} report={false} />
<div>{submitButton} {lockedInfo}</div><div/>
case "report":
  return <Runtime authoredState={authoredState} interactiveState={interactiveState} setInteractiveState={setInteractiveState} report={true} />
default:
  return "Loading...";
```
It is a little redundant but it seems more clear. To reduce the redundancy the runtime and report body could be pulled out to a function that takes a report variable so then it'd just be:
```
switch(mode){
case "authoring": 
  if(Authoring)....
case "runtime":
  return renderRuntime({report:false})
case "report":
  return renderRuntime({report:true})
default:
  return "Loading...";
```